frontend/src/components/editor/app-container.tsx

```diff
@@ -20,6 +27,52 @@ export const AppContainer: React.FC<PropsWithChildren<Props>> = ({
   isRunning,
   children,
 }) => {
+  const errors = useCellErrors();
+
+  // Dynamically update favicon for run feedback
+  let favicon: HTMLLinkElement | null =
+    document.querySelector("link[rel~='icon']");
+
+  if (!favicon) {
+    favicon = document.createElement("link");
+    favicon.rel = "icon";
+    document.getElementsByTagName("head")[0].append(favicon);
+  }
+
+  const favicons: Record<string, string> = useMemo(
```

Comment:
this doesnt need live in the react component. 

it could be at the module level:
```ts
const FAVICON_ROUTES = {
    idle: "./favicon.ico",
    success: "./circle-check.ico",
    running: "./circle-play.ico",
    error: "./circle-x.ico",
}
```